### PR TITLE
Fix multi gpu rendering

### PR DESCRIPTION
**Pull request by AlexanderVeselov**

### CODE REVIEW

File: Baikal/SceneGraph/scene_object.cpp [hunk 1]
```diff
@@ -1,16 +1,64 @@
...
+    }
+
     void SceneObject::ResetId()
     {
         g_next_id = 0;
     }
+
+    void SceneObject::SetSceneControllerId(std::uint32_t controller_id)
+    {
+        assert(controller_id >= 0 && controller_id < sizeof(std::uint32_t) * 8);
```
Comment (den13k) [inline Baikal/SceneGraph/scene_object.cpp]: the first one is always true.
pls. remove

File: Baikal/SceneGraph/scene_object.cpp [hunk 2]
```diff
@@ -1,16 +1,64 @@
...
     SceneObject::SceneObject()
-        : m_dirty(false), m_id(g_next_id++)
+        : m_dirty(0), m_id(g_next_id++)
     {
     }
 
+    bool SceneObject::IsDirty() const
+    {
+        bool dirty = false;
+        if (g_scene_controller_id == -1)
```
Comment (den13k) [inline Baikal/SceneGraph/scene_object.cpp]: is this possible?
assert should be enough

File: Baikal/SceneGraph/scene_object.cpp [hunk 3]
```diff
@@ -1,16 +1,64 @@
...
+
+    void SceneObject::SetDirty(bool dirty) const
+    {
+        if (dirty)
+        {
+            m_dirty = 0xFFFFFFFF;
+        }
+        else
+        {
+            if (g_scene_controller_id == -1)
```
Comment (den13k) [inline Baikal/SceneGraph/scene_object.cpp]: is this possible?

File: Baikal/Controllers/scene_controller.inl [hunk 4]
```diff
@@ -16,9 +16,13 @@
 
 namespace Baikal
 {
+    static std::uint32_t g_next_scene_controller_id = 0;
+
     template <typename CompiledScene>
     inline
-    SceneController<CompiledScene>::SceneController() {}
+    SceneController<CompiledScene>::SceneController()
+        : m_id(g_next_scene_controller_id++)
```
Comment (den13k) [inline Baikal/Controllers/scene_controller.inl]: does scene controllers persist while app is running? Is it possible that a controller will be deleted and the created with another id?

File: Baikal/SceneGraph/scene_object.cpp [hunk 5]
```diff
@@ -1,16 +1,64 @@
...
+    bool SceneObject::IsDirty() const
+    {
+        bool dirty = false;
+        if (g_scene_controller_id == -1)
+        {
+            dirty = (m_dirty != 0);
+        }
+        else
+        {
+            dirty = ((int)m_dirty & (1 << g_scene_controller_id)) == (1 << g_scene_controller_id);
```
Comment (den13k) [inline Baikal/SceneGraph/scene_object.cpp]: pls. introduce an additional variable and simplify the statement.

File: Baikal/SceneGraph/scene_object.cpp [hunk 6]
```diff
@@ -1,16 +1,64 @@
...
+        }
+        else
+        {
+            if (g_scene_controller_id == -1)
+            {
+                m_dirty = 0;
+            }
+            else
+            {
+                m_dirty &= ~(1 << g_scene_controller_id);
```
Comment (den13k) [inline Baikal/SceneGraph/scene_object.cpp]: pls. add a comment

**Review by den13k — CHANGES_REQUESTED** [review]

**Comment by den13k** [comment]
please consider to use std::bitset

File: BaikalStandalone/Application/cl_render.cpp [hunk 7]
```diff
@@ -54,23 +54,14 @@ namespace Baikal
...
-                settings.platform_index,
-                settings.device_index);
-        }
-        catch (CLWException &)
-        {
-            force_disable_itnerop = true;
-            ConfigManager::CreateConfigs(settings.mode, false, m_cfgs, settings.num_bounces, settings.platform_index, settings.device_index);
-        }
+        // Create cl context
+        ConfigManager::CreateConfigs(
```
Comment (den13k) [inline BaikalStandalone/Application/cl_render.cpp:58]: why there is no exception handling anymore?

File: Baikal/SceneGraph/scene_object.h [hunk 8]
```diff
@@ -58,13 +59,16 @@ namespace Baikal
...
+        static void SetSceneControllerId(std::uint32_t controller_id);
+        static void ResetSceneControllerId();
 
     protected:
         // Constructor
         SceneObject();
         
     private:
-        mutable bool m_dirty;
+        static const int kMaxDirtyBits = 32;
```
Comment (den13k) [inline Baikal/SceneGraph/scene_object.h]: pls add a comment here about the value

File: Baikal/Controllers/scene_controller.inl (hunk 4 above)
Comment (den13k) [reply Baikal/Controllers/scene_controller.inl]: discussed

File: BaikalStandalone/Application/cl_render.cpp (hunk 7 above)
Comment (den13k) [reply BaikalStandalone/Application/cl_render.cpp:58]: discussed

File: BaikalStandalone/Utils/config_manager.cpp [hunk 9]
```diff
@@ -94,6 +94,7 @@ void ConfigManager::CreateConfigs(
 
             Config cfg;
             cfg.caninterop = false;
+            bool create_secondary = true;
```
Comment (den13k) [inline BaikalStandalone/Utils/config_manager.cpp]: pls. rename to createWithoutInterop or something similar

File: Baikal/SceneGraph/scene_object.h (hunk 8 above)
Comment (AlexanderVeselov) [reply Baikal/SceneGraph/scene_object.h]: Done

File: BaikalStandalone/Utils/config_manager.cpp (hunk 9 above)
Comment (AlexanderVeselov) [reply BaikalStandalone/Utils/config_manager.cpp]: Done

**Review by den13k — APPROVED** [review]

**Pull request closed without merging**